readme.md
# Laravel OAuth Client

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Total Downloads][ico-downloads]][link-downloads]
[![StyleCI][ico-styleci]][link-styleci]

Multi-purpose OAuth client, configurable for a sleuth of providers through the config file.

## Installation

Via Composer

``` bash
$ composer require werk365/laraveloauthclient
```

## Usage

Publish the config
```bash
$ php artisan vendor:publish --provider="Werk365\LaravelOAuthClient\LaravelOAuthClientServiceProvider"
```
Configure the config using the provided example.

Then use the package as follows:
```php
use Werk365\LaravelOAuthClient\LaravelOAuthClient as OAuth;

// ...

$oauth = new OAuth("vendorname");

// Returns array defined in config
$response = $oauth->getToken($code);

//Returns array defined in config
$response = $oauth->getInfo($accesstoken);

//Returns array defined in config
$response = $oauth->refreshToken($refreshtoken);
```

## Config
Example of config provided, edit values to match vendor spec:
```php
<?php

return [
    'vendor1' => [
        'client_id' => '1234',
        'client_secret' => '12345',
        'redirect_uri' => 'https://www.example.com/oauth2/vendor1',
        'token' => [
            'url' => 'https://login.vendor.example.com/oauth2/token',
            'method' => 'POST',
            'grant_type' => 'authorization_code',
            'fields' => [
                'access_token' => 'access_token',
                'expires_in' => 'expires_in',
                'refresh_token' => 'refresh_token',
            ],
            'auth' => 'body',
        ],
        'refresh' => [
            'url' => 'https://login.vendor.example.com/oauth2/token',
            'method' => 'POST',
            'grant_type' => 'authorization_code',
            'fields' => "*",
            'auth' => 'body',
        ],
        'info' => [
            'url' => 'https://login.vendor.example.com/oauth2/metadata',
            'method' => 'GET',
            'fields' => [
                'metadata1',
                'metadata2',
            ],
        ],
    ],
    'vendor2' => ['...'],
];

```

## Change log

Please see the [changelog](changelog.md) for more information on what has changed recently.

## Testing

``` bash
$ composer test
```

## Security

If you discover any security related issues, please email author email instead of using the issue tracker.

[ico-version]: https://img.shields.io/packagist/v/werk365/laraveloauthclient.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/werk365/laraveloauthclient.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/werk365/laraveloauthclient/master.svg?style=flat-square
[ico-styleci]: https://styleci.io/repos/328643005/shield

[link-packagist]: https://packagist.org/packages/werk365/laraveloauthclient
[link-downloads]: https://packagist.org/packages/werk365/laraveloauthclient
[link-styleci]: https://styleci.io/repos/328643005
[link-author]: https://github.com/365Werk
[link-contributors]: ../../contributors


changelog.md
# Changelog

All notable changes to `LaravelOAuthClient` will be documented in this file.

## Version 1.0

### Added
- Everything


src/LaravelOAuthClient.php
<?php

namespace Werk365\LaravelOAuthClient;

use Exception;
use Illuminate\Support\Facades\Http;

class LaravelOAuthClient
{
    private $vendors;
    private $code;
    private $refresh_token;
    private $vendor;
    private $access_token;
    private $info_body;

    public function __construct($vendor = null)
    {
        $this->vendors = config('laraveloauthclient');
        $this->vendor = $vendor ?? null;
    }

    public function setVendor($vendor)
    {
        $this->vendor = $vendor;
    }

    public function setCode($code)
    {
        $this->code = $code;
    }

    public function setRefreshToken($refresh_token)
    {
        $this->refresh_token = $refresh_token;
    }

    public function setAccessToken($access_token)
    {
        $this->access_token = $access_token;
    }

    public function getToken($code = null): array
    {
        $this->code = $code ?? $this->code;
        if (! $this->code) {
            throw new Exception('No code set');
        } elseif (! $this->vendor) {
            throw new Exception('No vendor set');
        }
        $response = $this->tokenHandler('token');

        return $response;
    }

    public function refreshToken($refresh_token = null)
    {
        $this->refresh_token = $refresh_token ?? $this->refresh_token;
        if (! $this->refresh_token) {
            throw new Exception('No refresh token set');
        } elseif (! $this->vendor) {
            throw new Exception('No vendor set');
        }
        $response = $this->tokenHandler('refresh');

        return $response;
    }

    public function getInfo($access_token = null, $body = null)
    {
        $this->info_body = $body ?? $this->info_body ?? null;
        $this->access_token = $access_token ?? $this->access_token;
        if (! $this->access_token) {
            throw new Exception('No access token set');
        } elseif (! $this->vendor) {
            throw new Exception('No vendor set');
        }

        return $this->infoHandler();
    }

    private function tokenHandler(string $type): ?array
    {
        $body = [
            'client_id' => $this->vendors[$this->vendor]['client_id'],
            'client_secret' => $this->vendors[$this->vendor]['client_secret'],
            'redirect_uri' => $this->vendors[$this->vendor]['redirect_uri'],
            'grant_type' => $this->vendors[$this->vendor][$type]['grant_type'],
        ];

        if ($type == 'token') {
            $body['code'] = $this->code;
        } elseif ($type == 'refresh') {
            $body['refresh_token'] = $this->refresh_token;
        }

        $token = [];

        switch ($this->vendors[$this->vendor][$type]['auth']) {
            case 'body':
                if ($this->vendors[$this->vendor][$type]['method'] === 'POST') {
                    $token = Http::asForm()
                        ->post($this->vendors[$this->vendor][$type]['url'], $body)
                        ->json();
                } else {
                    $token = Http::get($this->vendors[$this->vendor][$type]['url'], $body)
                        ->json();
                }
                break;
            case 'basic':
                if ($this->vendors[$this->vendor][$type]['method'] === 'POST') {
                    $token = Http::withBasicAuth($this->vendors[$this->vendor]['client_id'], $this->vendors[$this->vendor]['client_secret'])
                        ->asForm()
                        ->post($this->vendors[$this->vendor][$type]['url'], $body)
                        ->json();
                } else {
                    $token = Http::withBasicAuth($this->vendors[$this->vendor]['client_id'], $this->vendors[$this->vendor]['client_secret'])
                        ->get($this->vendors[$this->vendor][$type]['url'], $body)
                        ->json();
                }
        }

        $fields = [];

        if (is_array($this->vendors[$this->vendor][$type]['fields'])) {
            foreach ($this->vendors[$this->vendor][$type]['fields'] as $key => $field) {
                $fields[$field] = $token[$field];
            }
        } elseif ($this->vendors[$this->vendor][$type]['fields'] == '*') {
            $fields = $token;
        }

        return $fields ?? null;
    }

    private function infoHandler(): ?array
    {
        if ($this->vendors[$this->vendor]['info']['method'] === 'POST') {
            $info = Http::withToken($this->access_token)->asForm()
                ->post($this->vendors[$this->vendor]['info']['url'], $this->info_body)
                ->json();
        } else {
            $info = Http::withToken($this->access_token)
                ->get($this->vendors[$this->vendor]['info']['url'], $this->info_body)
                ->json();
        }
        $fields = [];
        foreach ($this->vendors[$this->vendor]['info']['fields'] as $field) {
            $fields[$field] = $info[$field];
        }

        return $fields ?? null;
    }
}


src/LaravelOAuthClientServiceProvider.php
<?php

namespace Werk365\LaravelOAuthClient;

use Illuminate\Support\ServiceProvider;

class LaravelOAuthClientServiceProvider extends ServiceProvider
{
    /**
     * Perform post-registration booting of services.
     *
     * @return void
     */
    public function boot(): void
    {
        // Publishing is only necessary when using the CLI.
        if ($this->app->runningInConsole()) {
            $this->bootForConsole();
        }
    }

    /**
     * Register any package services.
     *
     * @return void
     */
    public function register(): void
    {
        $this->mergeConfigFrom(__DIR__.'/../config/laraveloauthclient.php', 'laraveloauthclient');

        // Register the service the package provides.
        $this->app->singleton('laraveloauthclient', function ($app) {
            return new LaravelOAuthClient;
        });
    }

    /**
     * Get the services provided by the provider.
     *
     * @return array
     */
    public function provides()
    {
        return ['laraveloauthclient'];
    }

    /**
     * Console-specific booting.
     *
     * @return void
     */
    protected function bootForConsole(): void
    {
        // Publishing the configuration file.
        $this->publishes([
            __DIR__.'/../config/laraveloauthclient.php' => config_path('laraveloauthclient.php'),
        ], 'laraveloauthclient.config');
    }
}


config/laraveloauthclient.php
<?php

return [
    'vendor1' => [
        'client_id' => '1234',
        'client_secret' => '12345',
        'redirect_uri' => 'https://www.example.com/oauth2/vendor1',
        'token' => [
            'url' => 'https://login.vendor.example.com/oauth2/token',
            'method' => 'POST',
            'grant_type' => 'authorization_code',
            'fields' => [
                'access_token' => 'access_token',
                'expires_in' => 'expires_in',
                'refresh_token' => 'refresh_token',
            ],
            'auth' => 'body',
        ],
        'refresh' => [
            'url' => 'https://login.vendor.example.com/oauth2/token',
            'method' => 'POST',
            'grant_type' => 'authorization_code',
            'fields' => [
                'access_token' => 'access_token',
                'expires_in' => 'expires_in',
                'refresh_token' => 'refresh_token',
            ],
            'auth' => 'body',
        ],
        'info' => [
            'url' => 'https://login.vendor.example.com/oauth2/metadata',
            'method' => 'GET',
            'fields' => [
                'metadata1',
                'metadata2',
            ],
        ],
    ],
    'vendor2' => ['...'],
];
